src/group/mls_group/create_commit.rs

```diff
@@ -94,14 +94,25 @@ impl MlsGroup {
             &self.interim_transcript_hash,
         )?;
 
+        // Calculate tree hash
         let tree_hash = provisional_tree.tree_hash();
-        // TODO #186: Implement extensions
+
+        // Check for group extensions and include them
+        let extensions: Vec<Box<dyn Extension>> = if self.use_ratchet_tree_extension {
+            vec![Box::new(RatchetTreeExtension::new(
+                provisional_tree.public_key_tree_copy(),
```

Comment:
I agree that would be a good optimisation for larger groups. Ultimately you need to clone the whole tree of course when you send out a Welcome message, but you shouldn't have to do it when you just apply a Commit.

Reply:
I looked into this again and I don't see an easy way to improve the performance here. The problem is that we need the full payload of the group extensions for the group context, which is just a serialisation of the payload. The byte vector is later used to derive the epoch secret. So in order to get rid of all duplication of the public tree, the epoch secret would have to be generated by iterating over the tree. Looking into that, it turns out that it leads to a number of issues:
 - serde cannot deserialise generic impls, meaning we would have tor write all deserialisation for all extensions by hand
 - we need to introduce lifetimes for all extensions (because this particular one now depends on the lifetime of a tree), which is complicated at best and potentially impossible

Given the above, I'd like to punt on making this more efficient for now. This only introduces an overhead when the ratchet tree extension is actually used, which by design is is incompatible with really large groups anyway.

Reply:
Thanks for looking into this! I see that it's complicated and I think we can just land this for now. I feel like we've now run into a couple of these issues that we likely want to revisit later to improve performance, so I've created #359 for this issue in particular and #358 to keep track of existing performance bottlenecks.